flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetRecordReader.java

```
@@ -0,0 +1,196 @@
+		}
+	}
+
+	public void close() throws  IOException {
+		if (reader != null) {
+			reader.close();
+		}
+	}
+
+	@CheckReturnValue(when = When.NEVER)
+	public T nextRecord() {
+		return currentValue;
+	}
+
+	public void seek(long syncedBlock) throws IOException {
+		PageReadStore pages = null;
+		while (syncedBlock > 0) {
+			pages = reader.readNextRowGroup();
+		}
+
+		recordReader = createRecordReader(pages);
+	}
+
+	private RecordReader<T> createRecordReader(PageReadStore pages) throws IOException {
+		if (pages == null) {
+			throw new IOException("Expecting more rows but reached last block. Read " + current + " out of " + total);
+		}
+		MessageColumnIO columnIO = columnIOFactory.getColumnIO(readSchema, fileSchema, strictTypeChecking);
+		// TODO (hpeter) enable filter later
+		RecordReader<T> recordReader = columnIO.getRecordReader(pages, recordMaterializer, FilterCompat.NOOP);
+		return recordReader;
+	}
+
+	public long getCurrentBlock() {
+		return currentBlock;
+	}
+
+	public boolean reachEnd() {
+		return current >= total;
+	}
+
+	public boolean hasNextRecord() throws IOException {
+		boolean recordFound = false;
+
+		while (!recordFound) {
+			// no more records left
+			if (current >= total) {
+				return false;
+			}
+
+			try {
+				checkRead();
+				current++;
+				try {
+					currentValue = recordReader.read();
+				} catch (RecordMaterializationException e) {
+					LOG.debug("skipping a corrupt record");
+					continue;
+				}
+
+				if (recordReader.shouldSkipCurrentRecord()) {
+					LOG.debug("skipping record");
+				}
+
+				if (currentValue == null) {
+					current = totalCountLoadedSoFar;
+					LOG.debug("filtered record reader reached end of block");
+					continue;
+				}
+
+				recordFound = true;
+				LOG.debug("read value: {}", currentValue);
+			} catch (RuntimeException e) {
+				LOG.error(String.format("Can not read value at %d in block %d in file %s",
+					current, currentBlock, reader.getFile()), e);
+				return false;
```

We might want to throw an exception to cancel the job depending on the configuration.

---

Agree. Added.